src/service/ampdoc-impl.js

```diff
@@ -441,6 +508,142 @@ export class AmpDoc {
   contains(node) {
     return this.getRootNode().contains(node);
   }
+
+  /**
+   * @param {!VisibilityState} visibilityState
+   * @restricted
+   */
+  overrideVisibilityState(visibilityState) {
+    if (this.visibilityStateOverride_ != visibilityState) {
+      this.visibilityStateOverride_ = visibilityState;
+      this.updateVisibilityState_();
+    }
+  }
+
+  /** @private */
+  updateVisibilityState_() {
+    // Natural visibility state.
+    const naturalVisibilityState = getDocumentVisibilityState(
+      this.win.document
+    );
+
+    // Parent visibility: pick the firts non-visible state.
+    let parentVisibilityState = VisibilityState.VISIBLE;
+    for (let p = this.parent_; p; p = p.getParent()) {
+      if (p.getVisibilityState() != VisibilityState.VISIBLE) {
+        parentVisibilityState = p.getVisibilityState();
+        break;
+      }
+    }
+
+    // Pick the most restricted visibility state.
+    let visibilityState;
+    const visibilityStateOverride =
+      this.visibilityStateOverride_ || VisibilityState.VISIBLE;
+    if (
+      naturalVisibilityState == VisibilityState.HIDDEN &&
+      visibilityStateOverride == VisibilityState.PAUSED
+    ) {
+      // Hidden document state overrides "paused".
+      visibilityState = naturalVisibilityState;
+    } else if (
+      visibilityStateOverride == VisibilityState.PAUSED ||
+      visibilityStateOverride == VisibilityState.INACTIVE
+    ) {
+      visibilityState = visibilityStateOverride;
+    } else if (
+      parentVisibilityState == VisibilityState.PAUSED ||
+      parentVisibilityState == VisibilityState.INACTIVE
+    ) {
+      visibilityState = parentVisibilityState;
+    } else if (
+      visibilityStateOverride == VisibilityState.PRERENDER ||
+      naturalVisibilityState == VisibilityState.PRERENDER ||
+      parentVisibilityState == VisibilityState.PRERENDER
+    ) {
+      visibilityState = VisibilityState.PRERENDER;
+    } else if (
+      visibilityStateOverride == VisibilityState.VISIBLE &&
+      parentVisibilityState == VisibilityState.VISIBLE &&
+      naturalVisibilityState == VisibilityState.VISIBLE
+    ) {
+      visibilityState = VisibilityState.VISIBLE;
+    } else {
+      visibilityState = VisibilityState.HIDDEN;
+    }
```

Comment:
a bit hard to read this block of if-else.

is it to pick most restricted state among `visibilityStateOverride` & `parentVisibilityState` & `naturalVisibilityState`? Can we sort of assign weight to each states and do a match calculation instead of if-else?

Also, might be helpful to have a good unit test to explain your intent.

Reply:
Tend to agree. But I moved this code mostly "as is" from the viewer-impl.js for cleaner refactoring. I reordered if-else here for (imho) better readability though. PTAL.